### PR TITLE
Implement UI for Admin Search of Hashtags

### PR DESCRIPTION
**Pull request by ThisIsMissEm**
This is a very rough first pass, but gives the following UI:

And with more than 20 tags:

image

Implements #29986, without the dynamic autocomplete for now.

### CODE REVIEW

**Comment by ThisIsMissEm** [comment]
@renchap do we have any tooling in the admin UI to upgrade a text input to a searchable text input? I don't really want to make this a react widget by default, but instead enhance it in some way?

File: app/views/admin/tags/index.html.haml
Comment (ThisIsMissEm) [inline app/views/admin/tags/index.html.haml:37]: Arguably maybe this partial should be moved here (`admin/tags`) and referenced from `admin/trends/tags`

The `f: nil` thing is simply to remove the checkboxes, since in this view we're not doing bulk actions right now, and tbh, I don't know if bulk actions would be desired here.

File: app/models/tag.rb
Comment (ThisIsMissEm) [inline app/models/tag.rb:137]: This change was necessary because in the Admin UI we want to see all hashtags when we search for a hashtag, so admin sets `include_unlistable` to true; I thought this was the cleaner way of handling this, instead of an `exclude_unlistable` and then trying to find all the code paths that used `search_for` to make them add that option.

File: spec/models/admin/tag_filter_spec.rb
Comment (ThisIsMissEm) [inline spec/models/admin/tag_filter_spec.rb:13]: This test should exercise the code change in `app/models/tag.rb` to allow searching outside of listable tags.

File: spec/models/tag_spec.rb
Comment (ThisIsMissEm) [inline spec/models/tag_spec.rb:249]: Arguably the limit / offset here should probably be via an option as well, but that involves modifying code well out of scope of this PR, hence leaving as is for now.

File: config/locales/en.yml
Comment (ThisIsMissEm) [inline config/locales/en.yml:899]:
```suggestion
      title: Hashtags
```

File: config/locales/en.yml
Comment (ThisIsMissEm) [inline config/locales/en.yml:883]:
```suggestion
      back_to_tags: Back to hashtags
```

**Comment by shleeable** [comment]
Request: Please add the ability to rename the case sensitivity if that's possible.

`#youtube` to `#YouTube` etc etc.

File: app/models/tag.rb
Comment (ClearlyClaire) [reply app/models/tag.rb:137]: Having something like `{ exclude_unlistable: true }.merge(options)` would allow a `exclude_unlistable` option without changing the call sites, but I think `include_unlistable` is fine too.

**Comment by ClearlyClaire** [comment]
Looks ok overall but I'd need to have a more careful review.

> Request: Please add the ability to rename the case sensitivity if that's possible.
> 
> `#youtube` to `#YouTube` etc etc.

I'm not sure what you mean, isn't that already possible?

**Comment by ThisIsMissEm** [comment]
> Request: Please add the ability to rename the case sensitivity if that's possible.
> 
> `#youtube` to `#YouTube` etc etc.

That's out of scope for this PR. But if there's an issue, maybe it'd be possible for me to work on it, but we'd have to work out how the rename would affect already published posts.

File: app/models/tag.rb
Comment (ThisIsMissEm) [reply app/models/tag.rb:137]: Oooh, that could be a good call. I think I like that better given there's already an exclude_... for something else here

File: app/models/admin/tag_filter.rb
Comment (oneiros) [inline app/models/admin/tag_filter.rb:20]: Why the `#to_s`? You did call `#symbolize_keys` on the params hash in the constructor, so keys should already be symbols. Or am I missing something?

File: app/models/admin/tag_filter.rb
Comment (oneiros) [inline app/models/admin/tag_filter.rb:31]: Same as above: Keys should be symbols, why do the extra step of converting them?

File: app/models/tag.rb
Comment (oneiros) [reply app/models/tag.rb:137]: Do you want to make that change then? Personally I believe both approaches are perfectly fine, but I agree it could make for a nicer "symmetry" with `exclude_unreviewed`.

File: app/views/admin/trends/tags/_tag.html.haml
Comment (oneiros) [inline app/views/admin/trends/tags/_tag.html.haml:2]: An alternative could be to use

```haml
- if local_assigns[:f]
```

I always found this to be quite ugly, but it would mean you could get rid of the `f: nil` argument when rendering this partial from `app/views/admin/tags/index.html.haml`.

**Comment by oneiros** [comment]
I left a couple of small remarks, but overall I think this is fantastic!

As for the suggestions to rename "tag" to "hashtag" in the locales, I agree with this change. I looked around the UI a bit and "hashtag" seems to be the wording commonly used. So for consistency's sake, I would agree to making these changes.

Lastly, as codecov is complaining, I believe it would be great to have some more examples in `spec/models/admin/tag_filter_spec.rb`.

File: app/models/admin/tag_filter.rb
Comment (ThisIsMissEm) [reply app/models/admin/tag_filter.rb:20]: This was based largely ob the AccountFilter, and I don't know enough about kaminari and our pagination helpers to say with confidence that a change would be okay

**Comment by ThisIsMissEm** [comment]
> Lastly, as codecov is complaining, I believe it would be great to have some more examples in spec/models/admin/tag_filter_spec.rb.

@oneiros yeah, I wasn't sure whether the filters were actually useful, since you're always searching for a tag pretty much.

So more looking for user feedback here: do tje filters make sense? Do we need them? Should we change them?

File: app/views/admin/trends/tags/_tag.html.haml
Comment (ThisIsMissEm) [reply app/views/admin/trends/tags/_tag.html.haml:2]: This seems like it could be nice, but I'm thinking I actually want to refactor this to not use that partial since I'm not sure how useful it is

File: app/models/tag.rb
Comment (ThisIsMissEm) [reply app/models/tag.rb:137]: Yeah, I will, have just been dealing with other things this week

File: app/models/admin/tag_filter.rb
Comment (oneiros) [reply app/models/admin/tag_filter.rb:20]: > This was based largely ob the AccountFilter

Ah, I see. Well, there is the same oddity there. But I see no need to repeat it here :slightly_smiling_face: 

> and I don't know enough about kaminari and our pagination helpers to say with confidence that a change would be okay

I do not believe pagination would be affected here. Pagination happens in the controller and uses the original params, not the ones from this class (that were converted to a hash with symbols as keys).

File: app/views/admin/trends/tags/_tag.html.haml
Comment (ThisIsMissEm) [reply app/views/admin/trends/tags/_tag.html.haml:2]: Have gone with a custom tag row partial, which emphasizes the moderation status as a whole, instead of emphasizing the trending status.

**Comment by ThisIsMissEm** [comment]
Here's the updated screenshots:

Screenshot 2024-07-05 at 23 55 46
Screenshot 2024-07-05 at 23 55 52

I've also dropped the "order by pending" since it didn't really make sense in practice. 

The trending hashtags UI uses an icon for the hash, I'm not sure why, but it looked rather weird with the new tag row partial, so I switched to just using a textual # instead.

File: config/locales/simple_form.en.yml
Comment (ThisIsMissEm) [inline config/locales/simple_form.en.yml:314]: See: https://github.com/mastodon/mastodon/blob/main/app/validators/disallowed_hashtags_validator.rb

File: app/javascript/styles/mastodon/accounts.scss
Comment (ThisIsMissEm) [inline app/javascript/styles/mastodon/accounts.scss:350]: I'm not 100% sold on this, but given all hashtags are unreviewed by default, the "pending review" became very overbearing in the list view with font-weight 500.

File: app/models/tag.rb
Comment (ThisIsMissEm) [inline app/models/tag.rb:136]: I could've used `merge!` here, but that felt kind of wrong for some reason

File: app/views/admin/tags/_tag.html.haml
Comment (ThisIsMissEm) [inline app/views/admin/tags/_tag.html.haml:3]: I really don't like this relying on the pending-account UI, since it's just generic batch-table contents stuff by now, but I'm also not thinking that refactoring that would at all be in scope of this PR.

File: app/views/admin/tags/_tag.html.haml
Comment (ThisIsMissEm) [inline app/views/admin/tags/_tag.html.haml:6]: Not sure using a translation here is correct, but it felt more correct than `= '#' + tag.display_name` or something like that.

File: config/locales/en.yml
Comment (ThisIsMissEm) [inline config/locales/en.yml:887]: There's a subtle difference between these two, arguably we could combine them, but I think we may want `Review requested` as a way to report tags as well in the future?

(Unless we change to polymorphic reports)

File: app/models/admin/tag_filter.rb
Comment (ThisIsMissEm) [inline app/models/admin/tag_filter.rb:44]: @oneiros can the rails params return a value that's **not** a string?

File: config/navigation.rb
Comment (ThisIsMissEm) [inline config/navigation.rb:43]: This change does mean that the `Trending > Hashtags` isn't highlighted when editing a tag from trending. The best we could do to get that back might be to use a query string or something?

```
/admin/tags/:id?trending=true
```

File: app/views/admin/tags/_tag.html.haml
Comment (ThisIsMissEm) [reply app/views/admin/tags/_tag.html.haml:6]: Interesting we have similar logic here: https://github.com/mastodon/mastodon/blob/74ead7d10698c7f18ca22c07d2f04ff81f419097/app/serializers/activitypub/hashtag_serializer.rb#L15

Perhaps it makes sense to have a `Tag#formatted_name` method?

File: app/models/tag.rb
Comment (oneiros) [reply app/models/tag.rb:136]: That is one reason `ActiveSupport` adds `#reverse_merge`. So you could do:

```ruby
options.reverse_merge!({ exclude_unlistable: true, exclude_unreviewed: false })
```

(I just wanted to provide an alternative, but do not have a strong preference myself.)

File: app/models/admin/tag_filter.rb
Comment (oneiros) [reply app/models/admin/tag_filter.rb:44]: Not in case of "regular" params sent by a browser. Since those are not typed, they will always be strings.

There is a special case, where a client can send JSON data and they end up in `params` as well. Since they are parsed as JSON, other data types may appear.

Of course this is not the case here. But there is another consideration: The controller could set a value before passing the params in here. I think here it already does to set a default order. One might want to allow symbols in this case, because they might look nicer in the controller. In that case `#to_s` would make sense here.

I would leave this up to you if you would like to allow this / guard against this here.

File: app/views/admin/tags/_tag.html.haml
Comment (oneiros) [reply app/views/admin/tags/_tag.html.haml:6]: I think it would!

And I would like that a lot better than the translation based approach.

Comment (ThisIsMissEm) [reply app/views/admin/tags/_tag.html.haml]: Okay, have implemented this, and also updated everywhere that we had this "##{...}" syntax for tags.

**Comment by ThisIsMissEm** [comment]
Okay, have updated this based on PR review feedback

File: app/serializers/activitypub/actor_serializer.rb
Comment (oneiros) [inline app/serializers/activitypub/actor_serializer.rb:189]: `Tag#formatted_name` uses the `#display_name` while this method used to use the `#name`. Those can be different (`name` being a normalized version of `display_name as per #18795).

So this here would actually change the behavior. (Of course I am in no way certain the old behavior was actually the intended one or an oversight in the linked PR :confused:).

Comment (ThisIsMissEm) [reply app/serializers/activitypub/actor_serializer.rb:189]: Hmm, good point! I'm not sure what the intended result should be here?

No tests began failing with this change 

cc @ClearlyClaire

Comment (ClearlyClaire) [reply app/serializers/activitypub/actor_serializer.rb:189]: That is a very good point. I suspect we might want `formatted_name` indeed, but I'd prefer that to be in another PR, as this does not seem related to the change at hand.

File: app/models/admin/tag_filter.rb
Comment (ClearlyClaire) [inline app/models/admin/tag_filter.rb:62]: It seems to me many of these filters are orthogonal with each other and thus could be split. But this can be changed later if we need to, this does not need to block the PR.

File: app/serializers/activitypub/note_serializer.rb
Comment (ClearlyClaire) [inline app/serializers/activitypub/note_serializer.rb:1]: This change should be in a different PR imho.

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]
This looks pretty good overall, I'd avoid the unrelated changes that I highlighted with inline comments, especially since some are more than refactoring and subtly change behavior, but otherwise this looks fine.

The filters could possibly be improved, but they seem useful already, and our moderators said they definitely see value in at least filtering out tags they already disabled.

**Comment by ThisIsMissEm** [comment]
@ClearlyClaire I only did the formatted_name change following discussion with @oneiros — https://github.com/mastodon/mastodon/pull/30880#discussion_r1667188572

**Comment by ClearlyClaire** [comment]
I can take over if you want.

**Comment by ThisIsMissEm** [comment]
> I can take over if you want.

I'll see if I can get to this today, but I've also a bit of freelance work todo (fixing an outage/bug) + the activitypub & IETF 120 meetings.

@ClearlyClaire to be clear, we're saying revert most of the Tag#formatted_name changes?

**Comment by ClearlyClaire** [comment]
> @ClearlyClaire to be clear, we're saying revert most of the Tag#formatted_name changes?

Yes. I'm fine with the refactor when there is no `name`/`display_name` change, what I want to avoid is the subtle behavior change. But I'm fine too with the `Tag#formatted_name` thing being left out of the PR completely. If you do leave it out, please use something like `"##{tag.display_name}"` rather than your initial i18n-based approach.

**Comment by ThisIsMissEm** [comment]
@ClearlyClaire I've rebased this PR and split out the other Tag#formatted_name changes into #31170

**Review by oneiros — APPROVED** [review]

**Review by ClearlyClaire — APPROVED** [review]

File: config/navigation.rb
Comment (ClearlyClaire) [reply config/navigation.rb:43]: This seems like something we'll need to address later, but I'm not sure how at the moment.

**Comment by ShadowJonathan** [comment]
Finally!! 🎉

**Comment by nileshtrivedi** [comment]
@ThisIsMissEm Thanks for this PR!

I have a somewhat related question. On my instance, I am trying to leverage hashtags to simulate Slack's channels. These can be shared-across-instance, curated-by-admin fixed set of topics that can enable organization and discovery of old posts and conversations. Conceptually they are instance-level featured tags.

Is there a way to accomplish this for all client apps without modifying Mastodon server code? I have seen the code for the [Tag model](https://github.com/mastodon/mastodon/blob/main/app/models/tag.rb) and this does not seem possible. Mastodon has [account-specific featured hashtags](https://docs.joinmastodon.org/user/discoverability/#featured-tags), but because most users on my instance use [Elk client](https://github.com/elk-zone/elk), this probably would not work.

**Comment by ThisIsMissEm** [comment]
@nileshtrivedi we currently don't have a notion of "recommended hashtags" though it has been suggested before I think, which would give you a page listing moderator recommended hashtags, which I think would accomplish what you want.

(it also makes sense following #31292) 

But hashtags still wouldn't be "instance specific" since all hashtags are global as far as I know

**Comment by nileshtrivedi** [comment]
Thanks @ThisIsMissEm . I came across #21772 and submitted this feature request in #31505 .

**Pull request merged**
